src/common/module_builder.cc

```diff
@@ -62,73 +55,152 @@ void ModuleBuilder::BuildModule(std::string_view code, std::string_view format,
 
   mlir::tt::registerAllDialects(registry);
   mlir::stablehlo::registerAllDialects(registry);
+
   mlir::func::registerAllExtensions(registry);
   mlir::tt::registerAllExtensions(registry);
 
-  context.appendDialectRegistry(registry);
+  mlir::tt::ttir::registerPasses();
+  mlir::tt::ttnn::registerPasses();
+
+  m_context->appendDialectRegistry(registry);
+}
+
+tt_pjrt_status ModuleBuilder::buildModule(const std::string_view &code,
+                                          const std::string_view &format) {
+  DLOG_F(LOG_DEBUG, "ModuleBuilder::buildModule");
+
+  mlir::OwningOpRef<mlir::ModuleOp> mlir_module = createVHLOModule(code);
+  if (!tt_pjrt_status_is_ok(m_status)) {
+    return m_status;
+  }
+
+  convertFromVHLOToSHLO(mlir_module);
+  if (!tt_pjrt_status_is_ok(m_status)) {
+    return m_status;
+  }
+
+  convertFromSHLOToTTIR(mlir_module);
+  if (!tt_pjrt_status_is_ok(m_status)) {
+    return m_status;
+  }
+
+  convertFromTTIRToTTNN(mlir_module);
+  if (!tt_pjrt_status_is_ok(m_status)) {
+    return m_status;
+  }
+
+  createFlatbufferBinary(mlir_module);
+
+  return m_status;
+}
 
-  mlir::OwningOpRef<mlir::ModuleOp> mlir_module =
+mlir::OwningOpRef<mlir::ModuleOp>
+ModuleBuilder::createVHLOModule(const std::string_view &code) {
+  mlir::OwningOpRef<mlir::ModuleOp> vhlo_module =
       mlir::parseSourceString<mlir::ModuleOp>(
           llvm::StringRef(code.data(), code.size()),
           // IR may be invalid because some fields may be using DenseElements
           // instead of DenseArray. We rectify that below and verify after.
-          mlir::ParserConfig{&context, /*verifyAfterParse=*/true});
-  DLOG_F(LOG_DEBUG, "VHLO Module");
-  if (log_level > 0)
-    mlir_module->dump();
+          mlir::ParserConfig{m_context.get(), /*verifyAfterParse=*/true});
 
-  mlir::PassManager vhlo_pm(mlir_module.get()->getName());
-  vhlo_pm.addPass(mlir::stablehlo::createVhloLegalizeToStablehloPass());
-  // Run the pass manager.
-  if (mlir::failed(vhlo_pm.run(mlir_module.get()))) {
-    throw std::runtime_error("Failed to run VHLO->SHLO pipeline.");
+  if (!vhlo_module) {
+    DLOG_F(ERROR, "Failed to create VHLO module from the input program code");
+    m_status = tt_pjrt_status::kInternal;
+    return nullptr;
   }
-  DLOG_F(LOG_DEBUG, "SHLO Module");
-  if (log_level > 0)
-    mlir_module->dump();
 
-  mlir::tt::ttir::registerPasses();
-  mlir::tt::ttnn::registerPasses();
+  DLOG_F(LOG_DEBUG, "VHLO Module:");
+  print_module(vhlo_module);
```

Comment:
Print always?

Reply:
It checks for verbosity inside the `print_module` function, to avoid having to do that in every place we want to print modules. There is potential for improvements here, for example we could move the log from line above into that function too, and we could have some `logger.h` file which has logging utilities so we are independent on the framework used for logging (like loguru), and move`LOG_DEBUG` there instead it being defined in `status.h`, etc. But I didn't want to tackle that in this PR, let's do it separately later.